emacs/config.org
#+TITLE: Gnu Emacs Config
#+AUTHOR: Mauro
#+STARTUP: showeverything
#+OPTIONS: toc:3

* TABLE OF CONTENTS :toc:
- [[#programs-to-load][PROGRAMS TO LOAD]]
  - [[#elpaca][elpaca]]
  - [[#load-evil-mode][load evil mode]]
  - [[#general-keybindings][general keybindings]]
  - [[#raibow-mode][raibow mode]]
  - [[#projectile][projectile]]
  - [[#diminish][diminish]]
  - [[#flycheck][Flycheck]]
  - [[#company][company]]
  - [[#dired][dired]]
- [[#ivy-autocomplete-tool][IVY autocomplete Tool]]
- [[#theme][THEME]]
- [[#fonts][FONTS]]
- [[#backups][BACKUPS]]
- [[#varia][VARIA]]
  - [[#dashboard-funcs][Dashboard funcs]]
  - [[#guiline-numbers][GUI+Line numbers]]
  - [[#language-support][LANGUAGE SUPPORT]]
  - [[#orgmodestuff][orgmodeStuff]]
  - [[#orgmode-indentation][orgmode indentation]]
  - [[#auto-blocks-mode][auto Blocks mode]]
  - [[#reload-initfile][reload Initfile]]
- [[#shells-and-terms][Shells and terms]]
  - [[#eshellactual-shell][Eshell(actual shell)]]
  - [[#vterm-terminal-emulator][vterm (terminal emulator)]]
  - [[#vterm-toggle][Vterm-toggle]]
- [[#sudo-edit][Sudo edit]]
- [[#all-the-icons][ALL THE ICONS]]
- [[#dashboard][DASHBOARD]]
- [[#applaucnher][APPLAUCNHER]]
  - [[#counsel-app-launcher][Counsel-App Launcher]]
- [[#buffer-move][BUFFER MOVE]]
- [[#git][GIT]]
  - [[#gittimemachine][GitTimeMachine]]
  - [[#magit][Magit]]
- [[#which-key][WHICH KEY]]

* PROGRAMS TO LOAD
** elpaca
#+begin_src emacs-lisp 
  (defvar elpaca-installer-version 0.11)
  (defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
  (defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
  (defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
  (defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                                :ref nil :depth 1 :inherit ignore
                                :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                                :build (:not elpaca--activate-package)))
  (let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
         (build (expand-file-name "elpaca/" elpaca-builds-directory))
         (order (cdr elpaca-order))
         (default-directory repo))
    (add-to-list 'load-path (if (file-exists-p build) build repo))
    (unless (file-exists-p repo)
      (make-directory repo t)
      (when (<= emacs-major-version 28) (require 'subr-x))
      (condition-case-unless-debug err
          (if-let* ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                    ((zerop (apply #'call-process `("git" nil ,buffer t "clone"
                                                    ,@(when-let* ((depth (plist-get order :depth)))
                                                        (list (format "--depth=%d" depth) "--no-single-branch"))
                                                    ,(plist-get order :repo) ,repo))))
                    ((zerop (call-process "git" nil buffer t "checkout"
                                          (or (plist-get order :ref) "--"))))
                    (emacs (concat invocation-directory invocation-name))
                    ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                          "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                    ((require 'elpaca))
                    ((elpaca-generate-autoloads "elpaca" repo)))
              (progn (message "%s" (buffer-string)) (kill-buffer buffer))
            (error "%s" (with-current-buffer buffer (buffer-string))))
        ((error) (warn "%s" err) (delete-directory repo 'recursive))))
    (unless (require 'elpaca-autoloads nil t)
      (require 'elpaca)
      (elpaca-generate-autoloads "elpaca" repo)
      (let ((load-source-file-function nil)) (load "./elpaca-autoloads"))))
  (add-hook 'after-init-hook #'elpaca-process-queues)
  (elpaca `(,@elpaca-order))
  (elpaca 'transient)



#+end_src

** load evil mode
#+begin_src emacs-lisp

  ;; Install a package via the elpaca macro
  ;; See the "recipes" section of the manual for more details.

  ;; (elpaca example-package)

  ;; Install use-package support
  (elpaca elpaca-use-package
    ;; Enable use-package :ensure support for Elpaca.
    (elpaca-use-package-mode))
  (setq use-package-always-ensure t)
  ;;When installing a package used in the init file itself,
  ;;e.g. a package which adds a use-package key word,
  ;;use the :wait recipe keyword to block until that package is installed/configured.
  ;;For example:
  ;;(use-package general :ensure (:wait t) :demand t)

  ;; Expands to: (elpaca evil (use-package evil :demand t))

  (use-package evil
    :init ;;tweaking config before loading it
    (setq evil-want-integration t)
    (setq evil-want-keybinding nil)
    (setq evil-vsplit-window-right t)
    (setq evil-undo-system 'undo-redo)
    (setq evil-split-window-below t)
    :config
    
    (evil-mode 1))

  (use-package evil-collection
    :after evil
    :init
    (setq evil-want-keybinding nil)
    :config
    
    (setq evil-collection-mode-list '(dashboard dired ibuffer))
    (evil-collection-init))

  (use-package evil-tutor
    :after evil)
  ;;Turns off elpaca-use-package-mode current declaration
  ;;Note this will cause evaluate the declaration immediately. It is not deferred.
  ;;Useful for configuring built-in emacs features.
  ;;(use-package emacs :ensure nil :config (setq ring-bell-function #'ignore))
#+end_src


** general keybindings
#+begin_src emacs-lisp
  (use-package general
    :config
    (general-evil-setup)
    ;; set up 'SPC' as the global leader key (space)
    (general-create-definer mleader-key
      :states '(normal insert visual emacs)
      :keymaps 'override
      :prefix "SPC" ;;set leader
      :global-prefix "M-SPC");;access leader in insert mode

    (mleader-key
      "g" '(:ignore t :wk "Git")    
      "g /" '(magit-displatch :wk "Magit dispatch")
      "g ." '(magit-file-displatch :wk "Magit file dispatch")
      "g b" '(magit-branch-checkout :wk "Switch branch")
      "g c" '(:ignore t :wk "Create") 
      "g c b" '(magit-branch-and-checkout :wk "Create branch and checkout")
      "g c c" '(magit-commit-create :wk "Create commit")
      "g c f" '(magit-commit-fixup :wk "Create fixup commit")
      "g C" '(magit-clone :wk "Clone repo")
      "g f" '(:ignore t :wk "Find") 
      "g f c" '(magit-show-commit :wk "Show commit")
      "g f f" '(magit-find-file :wk "Magit find file")
      "g f g" '(magit-find-git-config-file :wk "Find gitconfig file")
      "g F" '(magit-fetch :wk "Git fetch")
      "g g" '(magit-status :wk "Magit status")
      "g i" '(magit-init :wk "Initialize git repo")
      "g l" '(magit-log-buffer-file :wk "Magit buffer log")
      "g r" '(vc-revert :wk "Git revert file")
      "g s" '(magit-stage-file :wk "Git stage file")
      "g t" '(git-timemachine :wk "Git time machine")
      "g u" '(magit-stage-file :wk "Git unstage file"))

    (mleader-key
      "b" '(:ignore t :wk "buffer")
      "b b" '(switch-to-buffer :wk "Switch buffer")
      "b k" '(kill-this-buffer :wk "Kill this buffer")
      "b n" '(next-buffer :wk "Next buffer")
      "b p" '(previous-buffer :wk "Previous buffer")
      "b r" '(revert-buffer :wk "Reload buffer")
      "b i" '(ibuffer :wk "show all Buffers (iBuffer)"))
    (mleader-key
      "w" '(:ignore t :wk "Windows")
      ;; Window splits
      "w c" '(evil-window-delete :wk "Close window")
      "w n" '(evil-window-new :wk "New window")
      "w s" '(evil-window-split :wk "Horizontal split window")
      "w v" '(evil-window-vsplit :wk "Vertical split window")
      ;; Window motions
      "w h" '(evil-window-left :wk "Window left")
      "w j" '(evil-window-down :wk "Window down")
      "w k" '(evil-window-up :wk "Window up")
      "w l" '(evil-window-right :wk "Window right")
      "w w" '(evil-window-next :wk "Goto next window")
      ;; Move Windows
      "w H" '(buf-move-left :wk "Buffer move left")
      "w J" '(buf-move-down :wk "Buffer move down")
      "w K" '(buf-move-up :wk "Buffer move up")
      "w L" '(buf-move-right :wk "Buffer move right"))
    (mleader-key
      "d" '(:ignore t :wk "Dired")
      "d d" '(dired :wk "Open dired")
      "d j" '(dired-jump :wk "Dired jump to current")
      "d p" '(dired-preview-mode :wk "dired preview toggle"))
    (mleader-key
      "e" '(:ignore t :wk "eshell/evaluate")
      "e b" '(eval-buffer :wk "Evaluate elisp in buffer")
      "e d" '(eval-defun :wk "evaluate defun containing or after point");;define function
      "e e" '(eval-expression :wk "evaluate and elisp expression")
      "e l" '(eval-last-sexp :wk "evaluate elisp expression before point")
      "e s" '(eshell :wk "eshell")
      "e h" '(counsel-esh-history :wk "eshell history")
      "e r" '(eval-region :"evaluate elisp in region"))
    (mleader-key
      "SPC" '(counsel-M-x :wk "alt x")
      "." '(find-file :wk "open file")
      "f r" '(counsel-recentf :wk "find recent file")
      "TAB TAB" '(comment-line :wk "comment lines"))
    (mleader-key
      "h" '(:ignore t :wk "help")
      "h f" '(describe-function :wk "describe funtion")
      "h v" '(describe-variable :wk "describe variable")
      "h r r" '(reload-init-file :wk "reload config"))
    (mleader-key
      "t" '(:ignore :wk "toggle")
      "t l" '(display-line-numbers-mode :wk "toggle line numbers")
      "t t" '(visual-line-mode :wk "toggle truncated lines")
      "t v" '(vterm-toggle :wk "Toggle vterm"))
    )
#+end_src
** raibow mode
#+begin_src emacs-lisp
(use-package rainbow-mode
  :diminish
  :hook org-mode prog-mode)

#+end_src
** projectile
#+begin_src emacs-lisp
(use-package projectile
  :config
  (projectile-mode 1))

#+end_src
** diminish
#+begin_src emacs-lisp
(use-package diminish)
#+end_src

** Flycheck
#+begin_src emacs-lisp
;;this is a syntax checker because silly me can't type
(use-package flycheck
  :ensure t
  :defer t
  :diminish
  :init (global-flycheck-mode))
;;https://www.flycheck.org/en/latest/languages.html
#+end_src
** company
#+begin_src emacs-lisp
  (use-package company
    :defer 2
    :diminish
    :custom
    (company-begin-commands '(self-insert-command))
    (company-idle-delay .1)
    (company-minimum-prefix-length 2)
    (company-show-numbers t)
    (company-tooltip-align-annotations 't)
    (global-company-mode t))

  (use-package company-box
    :after company
    :diminish
    :hook (company-mode . company-box-mode))

#+end_src
** dired
#+begin_src emacs-lisp
    (use-package dired-open
      :config
      (setq dired-open-extensions '(("gif" . "sxiv")
                                    ("jpg" . "sxiv")
                                    ("png" . "sxiv")
                                    ("mkv" . "mpv")
                                    ("mp4" . "mpv"))))

    (use-package dired-preview
      :after dired
      :config
      (evil-define-key 'normal dired-mode-map
        (kbd "h") #'dired-up-directory
        (kbd "l") #'dired-find-file)
      :commands (dired-preview-mode)
      :config

      (setq dired-preview-trigger-on-start t
            dired-preview-delay 0.4
            dired-preview-max-size (* 10 (expt 2 20)) ; 10MB
            dired-preview-display-action-alist
            #'dired-preview-display-action-alist-below
            dired-preview-ignored-extensions-regexp
            "\\.\\(gz\\|zst\\|tar\\|xz\\|rar\\|zip\\|iso\\|epub\\)$"))

  #+end_src





* IVY autocomplete Tool
#+begin_src emacs-lisp
(use-package counsel
  :after ivy
  :diminish
  :config (counsel-mode))

(use-package ivy
  ;; :bind
  ;; ivy-resume resumes the last Ivy-based completion.
  ;; (("C-c C-r" . ivy-resume)
  ;;  ("C-x B" . ivy-switch-buffer-other-window))
  
  :diminish
  :custom
  (setq ivy-use-virtual-buffers t)
  (setq ivy-count-format "(%d/%d) ")
  (setq enable-recursive-minibuffers t)
  :config
  (ivy-mode))

(use-package all-the-icons-ivy-rich
  :ensure t
  :init (all-the-icons-ivy-rich-mode 1))

(use-package ivy-rich
  :after ivy
  :ensure t
  :init (ivy-rich-mode 1) ;; this gets us descriptions in M-x.
  :custom
  (ivy-virtual-abbreviate 'full
			  ivy-rich-switch-buffer-align-virtual-buffer t
			  ivy-rich-path-style 'abbrev)
  :config
  (ivy-set-display-transformer 'ivy-switch-buffer
                               'ivy-rich-switch-buffer-transformer))

#+end_src
* THEME
#+begin_src emacs-lisp
  (use-package doom-themes
    :config
    (setq doom-themes-enable-bold t
      	doom-themes-enable-italic t)
    (load-theme 'doom-zenburn  t)
    
    (set-face-attribute 'font-lock-comment-face nil :slant 'italic)
    (set-face-attribute 'font-lock-keyword-face nil :slant 'italic)

    )
  ;;zenburn
  ;;material
  (add-to-list 'custom-theme-load-path "~/.config/emacs/themes/")
  ;;(load-theme 'third t)
  #+end_src

* FONTS
Definining all Fonts emacs will use

#+begin_src emacs-lisp
  (set-face-attribute 'default nil
    :font "Iosevka"
    :height 110
    :weight 'medium)
  (set-face-attribute 'variable-pitch nil
    :font "Cantarell"
    :height 120
    :weight 'medium)
  (set-face-attribute 'fixed-pitch nil
    :font "Iosevka"
    :height 110
    :weight 'medium)

emacs/config.el
(defvar elpaca-installer-version 0.11)
(defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
(defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
(defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
(defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                              :ref nil :depth 1 :inherit ignore
                              :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                              :build (:not elpaca--activate-package)))
(let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
       (build (expand-file-name "elpaca/" elpaca-builds-directory))
       (order (cdr elpaca-order))
       (default-directory repo))
  (add-to-list 'load-path (if (file-exists-p build) build repo))
  (unless (file-exists-p repo)
    (make-directory repo t)
    (when (<= emacs-major-version 28) (require 'subr-x))
    (condition-case-unless-debug err
        (if-let* ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                  ((zerop (apply #'call-process `("git" nil ,buffer t "clone"
                                                  ,@(when-let* ((depth (plist-get order :depth)))
                                                      (list (format "--depth=%d" depth) "--no-single-branch"))
                                                  ,(plist-get order :repo) ,repo))))
                  ((zerop (call-process "git" nil buffer t "checkout"
                                        (or (plist-get order :ref) "--"))))
                  (emacs (concat invocation-directory invocation-name))
                  ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                        "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                  ((require 'elpaca))
                  ((elpaca-generate-autoloads "elpaca" repo)))
            (progn (message "%s" (buffer-string)) (kill-buffer buffer))
          (error "%s" (with-current-buffer buffer (buffer-string))))
      ((error) (warn "%s" err) (delete-directory repo 'recursive))))
  (unless (require 'elpaca-autoloads nil t)
    (require 'elpaca)
    (elpaca-generate-autoloads "elpaca" repo)
    (let ((load-source-file-function nil)) (load "./elpaca-autoloads"))))
(add-hook 'after-init-hook #'elpaca-process-queues)
(elpaca `(,@elpaca-order))
(elpaca 'transient)

;; Install a package via the elpaca macro
;; See the "recipes" section of the manual for more details.

;; (elpaca example-package)

;; Install use-package support
(elpaca elpaca-use-package
  ;; Enable use-package :ensure support for Elpaca.
  (elpaca-use-package-mode))
(setq use-package-always-ensure t)
;;When installing a package used in the init file itself,
;;e.g. a package which adds a use-package key word,
;;use the :wait recipe keyword to block until that package is installed/configured.
;;For example:
;;(use-package general :ensure (:wait t) :demand t)

;; Expands to: (elpaca evil (use-package evil :demand t))

(use-package evil
  :init ;;tweaking config before loading it
  (setq evil-want-integration t)
  (setq evil-want-keybinding nil)
  (setq evil-vsplit-window-right t)
  (setq evil-undo-system 'undo-redo)
  (setq evil-split-window-below t)
  :config
  
  (evil-mode 1))

(use-package evil-collection
  :after evil
  :init
  (setq evil-want-keybinding nil)
  :config
  
  (setq evil-collection-mode-list '(dashboard dired ibuffer))
  (evil-collection-init))

(use-package evil-tutor
  :after evil)
;;Turns off elpaca-use-package-mode current declaration
;;Note this will cause evaluate the declaration immediately. It is not deferred.
;;Useful for configuring built-in emacs features.
;;(use-package emacs :ensure nil :config (setq ring-bell-function #'ignore))

(use-package general
  :config
  (general-evil-setup)
  ;; set up 'SPC' as the global leader key (space)
  (general-create-definer mleader-key
    :states '(normal insert visual emacs)
    :keymaps 'override
    :prefix "SPC" ;;set leader
    :global-prefix "M-SPC");;access leader in insert mode

  (mleader-key
    "g" '(:ignore t :wk "Git")    
    "g /" '(magit-displatch :wk "Magit dispatch")
    "g ." '(magit-file-displatch :wk "Magit file dispatch")
    "g b" '(magit-branch-checkout :wk "Switch branch")
    "g c" '(:ignore t :wk "Create") 
    "g c b" '(magit-branch-and-checkout :wk "Create branch and checkout")
    "g c c" '(magit-commit-create :wk "Create commit")
    "g c f" '(magit-commit-fixup :wk "Create fixup commit")
    "g C" '(magit-clone :wk "Clone repo")
    "g f" '(:ignore t :wk "Find") 
    "g f c" '(magit-show-commit :wk "Show commit")
    "g f f" '(magit-find-file :wk "Magit find file")
    "g f g" '(magit-find-git-config-file :wk "Find gitconfig file")
    "g F" '(magit-fetch :wk "Git fetch")
    "g g" '(magit-status :wk "Magit status")
    "g i" '(magit-init :wk "Initialize git repo")
    "g l" '(magit-log-buffer-file :wk "Magit buffer log")
    "g r" '(vc-revert :wk "Git revert file")
    "g s" '(magit-stage-file :wk "Git stage file")
    "g t" '(git-timemachine :wk "Git time machine")
    "g u" '(magit-stage-file :wk "Git unstage file"))

  (mleader-key
    "b" '(:ignore t :wk "buffer")
    "b b" '(switch-to-buffer :wk "Switch buffer")
    "b k" '(kill-this-buffer :wk "Kill this buffer")
    "b n" '(next-buffer :wk "Next buffer")
    "b p" '(previous-buffer :wk "Previous buffer")
    "b r" '(revert-buffer :wk "Reload buffer")
    "b i" '(ibuffer :wk "show all Buffers (iBuffer)"))
  (mleader-key
    "w" '(:ignore t :wk "Windows")
    ;; Window splits
    "w c" '(evil-window-delete :wk "Close window")
    "w n" '(evil-window-new :wk "New window")
    "w s" '(evil-window-split :wk "Horizontal split window")
    "w v" '(evil-window-vsplit :wk "Vertical split window")
    ;; Window motions
    "w h" '(evil-window-left :wk "Window left")
    "w j" '(evil-window-down :wk "Window down")
    "w k" '(evil-window-up :wk "Window up")
    "w l" '(evil-window-right :wk "Window right")
    "w w" '(evil-window-next :wk "Goto next window")
    ;; Move Windows
    "w H" '(buf-move-left :wk "Buffer move left")
    "w J" '(buf-move-down :wk "Buffer move down")
    "w K" '(buf-move-up :wk "Buffer move up")
    "w L" '(buf-move-right :wk "Buffer move right"))
  (mleader-key
    "d" '(:ignore t :wk "Dired")
    "d d" '(dired :wk "Open dired")
    "d j" '(dired-jump :wk "Dired jump to current")
    "d p" '(dired-preview-mode :wk "dired preview toggle"))
  (mleader-key
    "e" '(:ignore t :wk "eshell/evaluate")
    "e b" '(eval-buffer :wk "Evaluate elisp in buffer")
    "e d" '(eval-defun :wk "evaluate defun containing or after point");;define function
    "e e" '(eval-expression :wk "evaluate and elisp expression")
    "e l" '(eval-last-sexp :wk "evaluate elisp expression before point")
    "e s" '(eshell :wk "eshell")
    "e h" '(counsel-esh-history :wk "eshell history")
    "e r" '(eval-region :"evaluate elisp in region"))
  (mleader-key
    "SPC" '(counsel-M-x :wk "alt x")
    "." '(find-file :wk "open file")
    "f r" '(counsel-recentf :wk "find recent file")
    "TAB TAB" '(comment-line :wk "comment lines"))
  (mleader-key
    "h" '(:ignore t :wk "help")
    "h f" '(describe-function :wk "describe funtion")
    "h v" '(describe-variable :wk "describe variable")
    "h r r" '(reload-init-file :wk "reload config"))
  (mleader-key
    "t" '(:ignore :wk "toggle")
    "t l" '(display-line-numbers-mode :wk "toggle line numbers")
    "t t" '(visual-line-mode :wk "toggle truncated lines")
    "t v" '(vterm-toggle :wk "Toggle vterm"))
  )

(use-package rainbow-mode
  :diminish
  :hook org-mode prog-mode)

(use-package projectile
  :config
  (projectile-mode 1))

(use-package diminish)

;;this is a syntax checker because silly me can't type
(use-package flycheck
  :ensure t
  :defer t
  :diminish
  :init (global-flycheck-mode))
;;https://www.flycheck.org/en/latest/languages.html

(use-package company
  :defer 2
  :diminish
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay .1)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  (global-company-mode t))

(use-package company-box
  :after company
  :diminish
  :hook (company-mode . company-box-mode))

(use-package dired-open
  :config
  (setq dired-open-extensions '(("gif" . "sxiv")
                                ("jpg" . "sxiv")
                                ("png" . "sxiv")
                                ("mkv" . "mpv")
                                ("mp4" . "mpv"))))

(use-package dired-preview
  :after dired
  :config
  (evil-define-key 'normal dired-mode-map
    (kbd "h") #'dired-up-directory
    (kbd "l") #'dired-find-file)
  :commands (dired-preview-mode)
  :config

  (setq dired-preview-trigger-on-start t
        dired-preview-delay 0.4
        dired-preview-max-size (* 10 (expt 2 20)) ; 10MB
        dired-preview-display-action-alist
        #'dired-preview-display-action-alist-below
        dired-preview-ignored-extensions-regexp
        "\\.\\(gz\\|zst\\|tar\\|xz\\|rar\\|zip\\|iso\\|epub\\)$"))

(use-package counsel
  :after ivy
  :diminish
  :config (counsel-mode))

(use-package ivy
  ;; :bind
  ;; ivy-resume resumes the last Ivy-based completion.
  ;; (("C-c C-r" . ivy-resume)
  ;;  ("C-x B" . ivy-switch-buffer-other-window))
  
  :diminish
  :custom
  (setq ivy-use-virtual-buffers t)
  (setq ivy-count-format "(%d/%d) ")
  (setq enable-recursive-minibuffers t)
  :config
  (ivy-mode))

(use-package all-the-icons-ivy-rich
  :ensure t
  :init (all-the-icons-ivy-rich-mode 1))

(use-package ivy-rich
  :after ivy
  :ensure t
  :init (ivy-rich-mode 1) ;; this gets us descriptions in M-x.
  :custom
  (ivy-virtual-abbreviate 'full
			  ivy-rich-switch-buffer-align-virtual-buffer t
			  ivy-rich-path-style 'abbrev)
  :config
  (ivy-set-display-transformer 'ivy-switch-buffer
                               'ivy-rich-switch-buffer-transformer))

(use-package doom-themes
  :config
  (setq doom-themes-enable-bold t
    	doom-themes-enable-italic t)
  (load-theme 'doom-zenburn  t)
  
  (set-face-attribute 'font-lock-comment-face nil :slant 'italic)
  (set-face-attribute 'font-lock-keyword-face nil :slant 'italic)

  )
;;zenburn
;;material
(add-to-list 'custom-theme-load-path "~/.config/emacs/themes/")
;;(load-theme 'third t)

(set-face-attribute 'default nil
  :font "Iosevka"
  :height 110
  :weight 'medium)
(set-face-attribute 'variable-pitch nil
  :font "Cantarell"
  :height 120
  :weight 'medium)
(set-face-attribute 'fixed-pitch nil
  :font "Iosevka"
  :height 110
  :weight 'medium)
;; Makes commented text and keywords italics.
;; This is working in emacsclient but not emacs.
;; Your font must have an italic face available.

;; This sets the default font on all graphical frames created after restarting Emacs.
;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
;; are not right unless I also add this method of setting the default font.
(add-to-list 'default-frame-alist '(font . "Iosevka-11"))    
;; Uncomment the following line if line spacing needs adjusting.
(setq-default line-spacing 0.12)

;;begone LOL
(setq backup-directory-alist '((".*" . "~/.local/share/Trash/files")))

(defun dashboard-greeting ()
  "Return a friendly greeting depending on time of day."
  (let ((hour (string-to-number (format-time-string "%H"))))
    (cond
     ((< hour 6)  "Still awake? 🌙")
     ((< hour 12) "Good morning! ☀️")
     ((< hour 18) "Good afternoon! ☕")
     (t           "Good evening! 🌆"))))

(delete-selection-mode 1)
(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)
(global-display-line-numbers-mode 1)
(global-visual-line-mode t)

;;add whatever language you need here
(use-package lua-mode)
(use-package haskell-mode)

;;table of content
(use-package toc-org
  :commands toc-org-enable
  :init
  (add-hook 'org-mode-hook 'toc-org-enable))
;;bullets
(add-hook 'org-mode-hook 'org-indent-mode)

(use-package org-bullets
  :config
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1))))

;;disables really strange indent behaviour
(electric-indent-mode -1)

(require 'org-tempo)

(defun reload-init-file()
  (interactive)
  (load-file user-init-file)
  (load-file user-init-file))

(use-package eshell-syntax-highlighting
    :after eshell
    :config
    (eshell-syntax-highlighting-global-mode +1))

(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands '("bash" "fish" "htop" "ssh" "top" "zsh"))

(use-package vterm
  :config
  (setq shell-file-name "/bin/fish"
	vterm-max-scrollback 5000))

(use-package vterm-toggle
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p nil)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
               '((lambda (buffer-or-name _)
                     (let ((buffer (get-buffer buffer-or-name)))
                       (with-current-buffer buffer
                         (or (equal major-mode 'vterm-mode)
                             (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
                  (display-buffer-reuse-window display-buffer-at-bottom)
                  ;;(display-buffer-reuse-window display-buffer-in-direction)
                  ;;display-buffer-in-direction/direction/dedicated is added in emacs27
                  ;;(direction . bottom)
                  ;;(dedicated . t) ;dedicated is supported in emacs27
                  (reusable-frames . visible)
                  (window-height . 0.25))))

(use-package sudo-edit
  :config
  (mleader-key 
    "f u" '(sudo-edit-find-file :wk "sudo find file")
    "f U" '(sudo-edit :wk "sudo edit file")))

(use-package all-the-icons
  :ensure t
  :if (display-graphic-p))

(use-package all-the-icons-dired
  :hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))

(use-package dashboard
  :ensure t 
  :init
  (setq initial-buffer-choice 'dashboard-open)
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title "Emacs Is More Than A Text Editor!")
  (setq dashboard-startup-banner "~/.config/emacs/images/Molly.jpeg") ;; use standard emacs logo as banner
  ;;(setq dashboard-startup-banner "/home/dt/.config/emacs/images/emacs-dash.png")  ;; use custom image as banner
  (setq dashboard-banner-logo-title (dashboard-greeting))
  (setq dashboard-center-content nil) ;; set to 't' for centered content
  (setq dashboard-items '((recents . 5)
                          (agenda . 5 )
                          (bookmarks . 3)
			  (projects . 3)
                          (registers . 3)))
  :custom
  (dashboard-modify-heading-icons '((recents . "file-text")
                                    (bookmarks . "book")))
  :config
  (dashboard-setup-startup-hook))

(defun emacs-counsel-launcher ()
  "Create and select a frame called emacs-counsel-launcher which consists only of a minibuffer and has specific dimensions. Runs counsel-linux-app on that frame, which is an emacs command that prompts you to select an app and open it in a dmenu like behaviour. Delete the frame after that command has exited"
  (interactive)
  (with-selected-frame 
    (make-frame '((name . "emacs-run-launcher")
                  (minibuffer . only)
                  (fullscreen . 0) ; no fullscreen
                  (undecorated . t) ; remove title bar
                  (auto-raise . t) ; focus on this frame
                  (tool-bar-lines . 0)
                  (menu-bar-lines . 0)
                  (internal-border-width . 10)
                  (width . 80)
                  (height . 11)))
                  (unwind-protect
                    (counsel-linux-app)
                    (delete-frame))))

(require 'windmove)

;;;###autoload
(defun buf-move-up ()
  "Swap the current buffer and the buffer above the split.
If there is no split, ie now window above the current one, an
error is signaled."
;;  "Switches between the current buffer, and the buffer above the
;;  split, if possible."'second
  (interactive)
  (let* ((other-win (windmove-find-other-window 'up))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No window above this one")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-down ()
"Swap the current buffer and the buffer under the split.
If there is no split, ie now window under the current one, an
error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'down))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (or (null other-win) 
            (string-match "^ \\*Minibuf" (buffer-name (window-buffer other-win))))
        (error "No window under this one")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-left ()
"Swap the current buffer and the buffer on the left of the split.
If there is no split, ie now window on the left of the current
one, an error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'left))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No left split")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

;;;###autoload
(defun buf-move-right ()
"Swap the current buffer and the buffer on the right of the split.
If there is no split, ie now window on the right of the current
one, an error is signaled."
  (interactive)
  (let* ((other-win (windmove-find-other-window 'right))
	 (buf-this-buf (window-buffer (selected-window))))
    (if (null other-win)
        (error "No right split")
      ;; swap top with this one
      (set-window-buffer (selected-window) (window-buffer other-win))
      ;; move this one to top
      (set-window-buffer other-win buf-this-buf)
      (select-window other-win))))

(use-package git-timemachine
  :hook (evil-normalize-keymaps . git-timemachine-hook)
  :config
  (evil-define-key 'normal git-timemachine-mode-map (kbd "C-j") 'git-timemachine-show-previous-revision)
  (evil-define-key 'normal git-timemachine-mode-map (kbd "C-k") 'git-timemachine-show-next-revision)
  )

(use-package magit)

(use-package which-key
:init
  (which-key-mode 1)
:config
(setq which-key-side-window-location 'bottom
	  which-key-sort-order #'which-key-key-order-alpha
	  which-key-sort-uppercase-first nil
	  which-key-add-column-padding 1
	  which-key-max-display-columns nil
	  which-key-min-display-lines 6
	  which-key-side-window-slot -10
	  which-key-side-window-max-height 0.25
	  which-key-idle-delay 0.8
	  which-key-max-description-length 25
	  which-key-allow-imprecise-window-fit nil
	  which-key-separator " → " ))
